### PR TITLE
Make COPY TO align with CREATE EXTERNAL TABLE

### PR DESCRIPTION
**Pull request by metesynnada**
## Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/9369.

## Rationale for this change

Changing 

```sql
OPTIONS (
    format X,
    X.foo.bar baz
)
``` 
into 
```sql
STORED AS X
OPTIONS (
    format.foo.bar baz
)
``` 
## What changes are included in this PR?

- "format" prefix is defined for all formats.
- COPY TO statement is now closer to CREATE EXTERNAL TABLE syntax.

More talk on https://github.com/synnada-ai/datafusion-upstream/pull/10

### CODE REVIEW

File: datafusion/sql/tests/sql_integration.rs [hunk 1]
```diff
@@ -397,7 +397,7 @@ CopyTo: format=csv output_url=output.csv options: ()
 
 #[test]
 fn plan_explain_copy_to() {
-    let sql = "EXPLAIN COPY test_decimal to 'output.csv'";
+    let sql = "EXPLAIN COPY test_decimal to 'output.csv' STORED AS CSV";
```
Comment (devinjdangelo) [inline datafusion/sql/tests/sql_integration.rs:400]: I think we should continue to support the original syntax here, which is in my opinion much more clear and concise.

File: datafusion/sqllogictest/test_files/clickbench.slt [hunk 2]
```diff
@@ -23,7 +23,7 @@
 # create.sql came from
 # https://github.com/ClickHouse/ClickBench/blob/8b9e3aa05ea18afa427f14909ddc678b8ef0d5e6/datafusion/create.sql
 # Data file made with DuckDB:
-# COPY (SELECT * FROM 'hits.parquet' LIMIT 10) TO 'clickbench_hits_10.parquet' (FORMAT PARQUET);
+# COPY (SELECT * FROM 'hits.parquet' LIMIT 10) TO 'clickbench_hits_10.parquet' STORED AS PARQUET;
```
Comment (devinjdangelo) [inline datafusion/sqllogictest/test_files/clickbench.slt:26]: Here I think we should also maintain backwards compatibility. It is nice to have a dedicated keyword for the often required format option, but I think user's queries in the old syntax should continue working

**Review by devinjdangelo — CHANGES_REQUESTED** [review]
Overall, I think the PR looks good, but I have some concerns about the syntax we are dropping. Most of all, I think queries like this should continue to work:

```
COPY table to 'file.csv'
```

Copy is useful on the commandline for quickly moving files around and this syntax is more compact and intuitive than

```
COPY table to 'file.csv' STORED AS CSV
```

So, I think we should continue supporting the ability to infer the format when copying to a single file with an unabiguous extension.

File: datafusion/sql/tests/sql_integration.rs (hunk 1 above)
Comment (metesynnada) [reply datafusion/sql/tests/sql_integration.rs:400]: I agree with this, I can made STORED AS optional.

File: datafusion/sqllogictest/test_files/clickbench.slt (hunk 2 above)
Comment (metesynnada) [reply datafusion/sqllogictest/test_files/clickbench.slt]: I firmly believe that a singular approach to the format introduction is the most logical and maintainable route. Thus implementing a breaking change aligns better with our goals for clarity and efficiency.

For this particular highlight, I didn't read the DuckDB reminder, this is why I changed it back.

Comment (alamb) [reply datafusion/sqllogictest/test_files/clickbench.slt]: I agree that supporting the duckdb style would be nice, but that this PR's consistency is also nice

I wonder if it would be possible (as a follow on PR) to add some special case backwards compatibility code to support the `(FORMAT PARQUET)` style? It seems to me like this could be a small amount of additional, localized code, and then we could going forward use the unified approach in this PR by default?

File: datafusion/sqllogictest/test_files/clickbench.slt [hunk 3]
```diff
@@ -23,7 +23,7 @@
 # create.sql came from
 # https://github.com/ClickHouse/ClickBench/blob/8b9e3aa05ea18afa427f14909ddc678b8ef0d5e6/datafusion/create.sql
 # Data file made with DuckDB:
-# COPY (SELECT * FROM 'hits.parquet' LIMIT 10) TO 'clickbench_hits_10.parquet' (FORMAT PARQUET);
+# COPY (SELECT * FROM 'hits.parquet' LIMIT 10) TO 'clickbench_hits_10.parquet';
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/clickbench.slt:26]: I think we should revert this change as it is a comment showing what command was used in duckdb (not a datafusion command)

**Review by alamb — COMMENTED** [review]
Thanks @metesynnada  -- other than the commented out test, this PR looks good to go in my opinion. 

It would be good to get @devinjdangelo  's opinion as well

I think the follow ups would be:
1. make the `format.` prefix optional
2. Support `(FORMAT PARQUET)` style option for backwards compatibility

**Comment by ozankabak** [comment]
> Thanks @metesynnada -- other than the commented out test, this PR looks good to go in my opinion.
> 
> It would be good to get @devinjdangelo 's opinion as well
> 
> I think the follow ups would be:
> 
> 1. make the `format.` prefix optional
> 2. Support `(FORMAT PARQUET)` style option for backwards compatibility

Sounds good. Let's get the consistent base syntax in with this PR, and have follow-ons for shortcuts (1) and backwards compatibility (2).

File: datafusion/sqllogictest/test_files/copy.slt [hunk 4]
```diff
@@ -111,49 +111,52 @@ a
...
-query TTT
-copy test to 'test_files/scratch/copy/escape_quote' (format csv, partition_by '''test2'',''test3''')
-----
-3
-
-statement ok
-CREATE EXTERNAL TABLE validate_partitioned_escape_quote STORED AS CSV 
-LOCATION 'test_files/scratch/copy/escape_quote/' PARTITIONED BY ("'test2'", "'test3'");
-
+## Until the partition by parsing uses ColumnDef, this test is meaningless since it becomes an overfit. Even in
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/copy.slt:114]: Here is the expalanation of why this test is commented out: https://github.com/synnada-ai/datafusion-upstream/pull/10#discussion_r1526407025

**Review by alamb — APPROVED** [review]
Thank you @metesynnada  -- I think this PR is good to go from my perspective after:
1. We file a ticket to track backwards compatible (`FORMAT PARQUET`) syntax
2. We file a ticket about what is going on with escaped quotes (aka https://github.com/synnada-ai/datafusion-upstream/pull/10#discussion_r1526407025)
3. @devinjdangelo  gives it a final review

File: datafusion/sqllogictest/test_files/copy.slt [hunk 5]
```diff
@@ -54,8 +54,8 @@ select * from validate_partitioned_parquet_bar order by col1;
 
 # Copy to directory as partitioned files
 query ITT
-COPY (values (1, 'a', 'x'), (2, 'b', 'y'), (3, 'c', 'z')) TO 'test_files/scratch/copy/partitioned_table2/' 
-(format parquet, partition_by 'column2, column3', 'parquet.compression' 'zstd(10)');
+COPY (values (1, 'a', 'x'), (2, 'b', 'y'), (3, 'c', 'z')) TO 'test_files/scratch/copy/partitioned_table2/' STORED AS parquet PARTITIONED BY (column2, column3)
```
Comment (devinjdangelo) [inline datafusion/sqllogictest/test_files/copy.slt:57]: This looks pretty cool. Definitely an improvement in readability over the prior syntax :+1:

**Review by devinjdangelo — APPROVED** [review]
Thank you @metesynnada this looks great! It definitely is nice now that copying data into files and then creating a table over the files have very similar syntaxes. And it remains easy/compact to copy to a single file, which is also very nice! :rocket: 

```
DataFusion CLI v36.0.0
❯ copy (values (1, 2, 3), (4, 5, 6)) to 'file.csv';
+-------+
| count |
+-------+
| 2     |
+-------+
1 row in set. Query took 0.020 seconds.

❯ select * from 'file.csv';
+---------+---------+---------+
| column1 | column2 | column3 |
+---------+---------+---------+
| 1       | 2       | 3       |
| 4       | 5       | 6       |
+---------+---------+---------+
2 rows in set. Query took 0.005 seconds.
❯ copy (values ('1', 2, 3), ('4', 5, 6)) to 'partitioned_csv/' partitioned by (column1) stored as csv;
+-------+
| count |
+-------+
| 2     |
+-------+
1 row in set. Query took 0.027 seconds.
❯ create external table partitioned_csv stored as CSV location 'partitioned_csv' partitioned by (column1) with header row;
0 rows in set. Query took 0.001 seconds.

❯ select * from partitioned_csv;
+---------+---------+---------+
| column2 | column3 | column1 |
+---------+---------+---------+
| 5       | 6       | 4       |
| 2       | 3       | 1       |
+---------+---------+---------+
2 rows in set. Query took 0.001 seconds.
```

File: datafusion/common/src/config.rs [hunk 6]
```diff
@@ -1628,38 +1616,15 @@ mod tests {
...
-        let mut table_config = TableOptions::new();
-        table_config
-            .set("parquet.bloom_filter_enabled::col1", "true")
-            .unwrap();
-        assert_eq!(
-            table_config.parquet.column_specific_options["col1"].bloom_filter_enabled,
-            Some(true)
-        );
-    }
-
```
Comment (ozankabak) [inline datafusion/common/src/config.rs:1642]: It is not obvious to me why this is removed and not its CSV counterpart. It seems this could be adapted like the CSV test too, as in:
```rust
    #[cfg(feature = "parquet")]
    #[test]
    fn parquet_table_options() {
        let mut table_config = TableOptions::new();
        table_config.set_file_format(FileType::PARQUET);
        table_config
            .set("format.bloom_filter_enabled::col1", "true")
            .unwrap();
        assert_eq!(
            table_config.parquet.column_specific_options["col1"].bloom_filter_enabled,
            Some(true)
        );
    }
```

File: datafusion/common/src/config.rs [hunk 7]
```diff
@@ -1628,38 +1616,15 @@ mod tests {
...
-    fn parquet_table_options_config_entry() {
-        let mut table_config = TableOptions::new();
-        table_config
-            .set("parquet.bloom_filter_enabled::col1", "true")
-            .unwrap();
-        let entries = table_config.entries();
-        assert!(entries
-            .iter()
-            .any(|item| item.key == "parquet.bloom_filter_enabled::col1"))
-    }
```
Comment (ozankabak) [inline datafusion/common/src/config.rs:1664]: Similar to my comment above

File: datafusion/sql/src/parser.rs [hunk 8]
```diff
@@ -111,15 +116,28 @@ impl fmt::Display for CopyToStatement {
...
+            write!(f, " PARTITIONED BY ({})", partitioned_by.join(", "))?;
+        }
+
+        if self.has_header {
+            write!(f, " WITH HEADER ROW")?;
+        }
 
         if !options.is_empty() {
             let opts: Vec<_> = options.iter().map(|(k, v)| format!("{k} {v}")).collect();
             // print them in sorted order
```
Comment (ozankabak) [inline datafusion/sql/src/parser.rs:139]: Is this comment accurate? I don't see an explicit sort here, is the `opts` vector somehow constructed in a sorted way?

**Review by ozankabak — APPROVED** [review]
LGTM, left three minor comments

File: datafusion/sql/src/parser.rs (hunk 8 above)
Comment (metesynnada) [reply datafusion/sql/src/parser.rs:139]: It was leftover.

**Comment by ozankabak** [comment]
This seems good to go - I will wait for a little bit more in case there are any more comments and then merge this today.

**Comment by ozankabak** [comment]
Let's address any remaining issues in quick follow-ons.

File: datafusion/sqllogictest/test_files/clickbench.slt (hunk 2 above)
Comment (alamb) [reply datafusion/sqllogictest/test_files/clickbench.slt]: Filed https://github.com/apache/arrow-datafusion/issues/9713 to track

File: datafusion/sqllogictest/test_files/copy.slt (hunk 4 above)
Comment (alamb) [reply datafusion/sqllogictest/test_files/copy.slt:114]: Filed https://github.com/apache/arrow-datafusion/issues/9714 to track

**Comment by alamb** [comment]
Filed https://github.com/apache/arrow-datafusion/issues/9716 to track the enhancement to avoid repeating `.format` over and ver

**Comment by alamb** [comment]
Thanks to @devinjdangelo  and @tinfoil-knight  I think we have completed all the follow on items https://github.com/apache/arrow-datafusion/issues/9716 and https://github.com/apache/arrow-datafusion/issues/9713



I filed https://github.com/apache/arrow-datafusion/pull/9754 to update the documentation
